File: VB/ClockTimeEditControl/RepositoryItemClockTimeEdit.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Drawing
Imports DevExpress.Utils
Imports DevExpress.XtraEditors.Drawing
Imports DevExpress.XtraEditors.Registrator
Imports DevExpress.XtraEditors.Repository

Namespace ClockTimeEditControl
	Public Class RepositoryItemClockTimeEdit
		Inherits RepositoryItemPictureEdit
		Private innerDateEditProperties As RepositoryItemDateEdit
		Private innerCalendar As ClockTimeCalendar
		Friend Const EditorName As String = "ClockTimeEdit"

		Shared Sub New()
			Register()
		End Sub
		Public Sub New()
			innerDateEditProperties = New RepositoryItemDateEdit()
			innerDateEditProperties.VistaEditTime = DefaultBoolean.True
			innerCalendar = New ClockTimeCalendar(innerDateEditProperties, DateTime.Now)
		End Sub

		Protected Overridable Function CalcClockSize() As Size
			Dim tempSize As New Size(1000, 1000)
			Dim bmp As New Bitmap(tempSize.Width, tempSize.Height)
			Dim g As Graphics = Graphics.FromImage(bmp)

			innerCalendar.ClientSize = tempSize
			Return innerCalendar.CalcBestSize(g)
		End Function

		Public Overridable Function ConvertTimeToClockImage(ByVal timeToConvert As DateTime) As Bitmap
			innerCalendar.DateTime = timeToConvert

			Dim calendarSize As Size = CalcClockSize()
			Dim bmp As New Bitmap(calendarSize.Width, calendarSize.Height)
			Dim graphics As Graphics = Graphics.FromImage(bmp)

			innerCalendar.ClientSize = calendarSize
			Dim args As New ClockTimeEditInfoArgs(innerCalendar)
			args.Graphics = graphics
			args.CalcLayout(New Rectangle(New Point(0, 0), calendarSize), Nothing)

			Dim painter As New VistaDateEditClockPainter(innerCalendar)
			painter.DrawObject(args)

			Return bmp
		End Function

		Public Shared Sub Register()
            EditorRegistrationInfo.Default.Editors.Add(New EditorClassInfo(EditorName, GetType(ClockTimeEdit), GetType(RepositoryItemClockTimeEdit), GetType(ClockTimeEditViewInfo), New PictureEditPainter(), True))
		End Sub

		Public Overrides ReadOnly Property EditorTypeName() As String
			Get
				Return EditorName
			End Get
		End Property
	End Class
End Namespace

File: CS/ClockTimeEditControl/RepositoryItemClockTimeEdit.cs
using System;
using System.Drawing;
using DevExpress.Utils;
using DevExpress.XtraEditors.Drawing;
using DevExpress.XtraEditors.Registrator;
using DevExpress.XtraEditors.Repository;

namespace ClockTimeEditControl
{
    public class RepositoryItemClockTimeEdit : RepositoryItemPictureEdit
    {
        private RepositoryItemDateEdit innerDateEditProperties;
        private ClockTimeCalendar innerCalendar;
        internal const string EditorName = "ClockTimeEdit";

        static RepositoryItemClockTimeEdit()
        {
            Register();
        }
        public RepositoryItemClockTimeEdit()
        {
            innerDateEditProperties = new RepositoryItemDateEdit();
            innerDateEditProperties.VistaEditTime = DefaultBoolean.True;
            innerCalendar = new ClockTimeCalendar(innerDateEditProperties, DateTime.Now);
        }

        protected virtual Size CalcClockSize()
        {
            Size tempSize = new Size(1000, 1000);
            Bitmap bmp = new Bitmap(tempSize.Width, tempSize.Height);
            Graphics g = Graphics.FromImage(bmp);

            innerCalendar.ClientSize = tempSize;
            return innerCalendar.CalcBestSize(g);
        }

        public virtual Bitmap ConvertTimeToClockImage(DateTime timeToConvert)
        {
            innerCalendar.DateTime = timeToConvert;

            Size calendarSize = CalcClockSize();
            Bitmap bmp = new Bitmap(calendarSize.Width, calendarSize.Height);
            Graphics graphics = Graphics.FromImage(bmp);

            innerCalendar.ClientSize = calendarSize;
            ClockTimeEditInfoArgs args = new ClockTimeEditInfoArgs(innerCalendar);
            args.Graphics = graphics;
            args.CalcLayout(new Rectangle(new Point(0, 0), calendarSize), null);

            VistaDateEditClockPainter painter = new VistaDateEditClockPainter(innerCalendar);
            painter.DrawObject(args);

            return bmp;
        }

        public static void Register()
        {
            EditorRegistrationInfo.Default.Editors.Add(new EditorClassInfo(EditorName, typeof(ClockTimeEdit),
                typeof(RepositoryItemClockTimeEdit), typeof(ClockTimeEditViewInfo),
                    new PictureEditPainter(), true, null));
        }

        public override string EditorTypeName
        {
            get { return EditorName; }
        }
    }
}
